_posts/2019-04-23-js-recon-is-no-more.md

```diff
@@ -0,0 +1,147 @@
+function I shared earlier. More than 5000 requests have ended in nearly no time.
+There were less than 1000 requests which ended in 1 microseconds. From the above
+results, We can conclude that mentioned method is not giving different results
+for occupied and empty port. We can conclude that mentioned method by AndLabs is
+failing to distinguish occupied ports from nonoccupied ports.
+
+I tried hard to find possible causes for the failure of this attack. I didn't
+find any conclusive evidence. Perhaps our hardware or browser code has improved
+for opening a TCP sockets quicker than what it used to. I will be honest, but it
+took sometime to understand the anatomy of this attack from that abstract blog
+post by Andlabs. However I was not about to give up and turn back from this
+point. Just for my satisfaction, I tried every possible combinations of
+`xhr.readyState` values to find any pattern. From my observation, I recognized
+that timings for returning a header from an occupied port was delayed.
+Comparatively, this response was quick for ports where no service was running. I
+am comparing the time in which the browser returned a response headers whereas
+in the previous method it was dependent on the time browser took for opening a
+socket.
+
+```javascript
+var requestPort = function(port) {
+  var startTime = null;
+  var xhr = new XMLHttpRequest();
+  xhr.onreadystatechange = function() {
+    if (xhr.readyState === 2) {
+      var timeTook = Date.now() - startTime;
+      console.log("Browser took : " + timeTook + " microseconds to send response headers.");
+    }
+  };
+  startTime = Date.now();
+  url = "http://localhost:" + port;
+  xhr.open("GET", url, true);
+  xhr.send(null);
+};
+```
+
+The above code will measure times the browser has taken to receive headers from
+the destination. You should call the function requestPort() with several
+combinations of empty and non-empty ports.
+
+![Graph showing header response time.]({{site.url}}/assets/images/js_recon_is_no_more/graph_header_response_time.png)
+
+This graph is representing header response times of 6000 requests fired at both
+active (27017) and inactive port (8084). The response times of inactive ports do
+not go beyond 200 microseconds. Comparing this with a header response time of
+occupied ports, we can see that response time of non-empty port is recognizably
+higher than empty port.
+
+The web browser is the most common tool we use. Asserting this vulnerability
+requires knowledge of Javascript and everyone is not a developer. As I mentioned
+earlier, I failed to find the source code of JS-Recon (the tool written by
+AndLabs proving possibility of this attack). For those reasons, I decided to
+write a tool pioneered on my improvements on an attempt of Andlabs.  Today, I
+have successfully completed that tool. I have decided to name it
+["Chatur"][chatur_pronounciation]. Chatur means intelligent person in
```

Comment:
Honestly? You could omit the explanation what it means as an easter egg, the ones who'll find out on their own will feel super smart...